lib/Sema/TypeCheckPattern.cpp

```diff
@@ -736,6 +736,74 @@ static bool validateParameterType(ParamDecl *decl, DeclContext *DC,
     }
     decl->getTypeLoc().setType(Ty);
   }
+  // If the param is not a 'let' and it is not an 'inout'.
+  // It must be a 'var'. Provide helpful diagnostics like a shadow copy
+  // in the function body to fix the 'var' attribute.
+  if (!decl->isLet() && !Ty->is<InOutType>()) {
```

Comment:
This whole blob of code is pretty detailed and separable for the rest of the logic in validateParameterType.  Please split the body of the if out to a new helper function, so we get something more like:

+  // If the param is not a 'let' and it is not an 'inout'.
  // It must be a 'var'. Provide helpful diagnostics like a shadow copy
  // in the function body to fix the 'var' attribute.
  if (!decl->isLet() && !Ty->is<InOutType>())
    diagnoseAndMigrateVarParameter(...)

or whatever.